api/src/main/java/com/epam/pipeline/manager/cloud/CloudFacadeImpl.java

```diff
@@ -259,6 +260,23 @@ public CloudInstanceState getInstanceState(final Long runId) {
         return getInstanceService(region).getInstanceState(region, String.valueOf(runId));
     }
 
+    @Override
+    public InstanceDNSRecord changeInstanceDNSRecord(Long regionId, final InstanceDNSRecord dnsRecord, final boolean delete) {
+
+        if (delete && regionId == null) {
+            List<PipelineRun> runs = pipelineRunManager.loadAllRunsByServiceURL(dnsRecord.getDnsRecord());
+            if (!runs.stream().allMatch(pipelineRun -> pipelineRun.getEndDate() != null)) {
+                log.warn("Won't try to delete dns record: " + dnsRecord.getDnsRecord() + " because it is still used a least in one active run.");
+                return null;
+            } else {
+                regionId = runs.stream().findFirst().map(pipelineRun -> pipelineRun.getInstance().getCloudRegionId()).orElse(null);
```

Comment:
Please, do not mutate input arguments, introduce a new variable instead. You can extract logic to determine actual region id to a separate method.